### ISSUE TITLE
Suggestion: Allow ReferenceResolvers to selectively assign reference IDs

### ISSUE THREAD

**Issue by mihtjel** [dotnet/runtime#94410]
We have some services returning somewhat complex objects with many sub-objects, including collections as well as references to specific elements in these collections. With the current `System.Text.Json.ReferenceResolver` implementation, I have to choose to either have `$id`-fields on all objects, or none. I would like to be able to only create them where needed, making the output terser.

I believe this arises from `ReferenceResolver.GetReference` not being allowed to return `null`, which is further enforced and relied upon in `JsonSerializer.Write.HandleMetadata.TryGetReferenceForValue(...)`.

As an example, I currently get this:

```
{
  "$id": "1",
  "id": 1234,
  "someObject": {
    "$id": "2",
    "someProperty": "value"
  },
  "someOtherObject": {
    "$id": "3",
    "someCollection": {
      "$id": "4",
      "$values": [
        {
          "$id": "5",
          ...
        },
        {
          "$id": "6",
          ...
        }
      ]
    },
    "primaryItem": {
      "$ref": "5"
    }
  }
}
```

What I really wanted in this case was more like this:

```
{
  "id": 1234,
  "someObject": {
    "someProperty": "value"
  },
  "someOtherObject": {
    "someCollection": [
        {
            "$id": "1",
            ...
        },
        {
            "$id": "2",
            ...
        }
    ],
    "primaryItem": {
      "$ref": "1"
    }
  }
}
```

I believe the changes required would be fairly limited, mostly in `ReferenceResolver.cs` and `JsonSerializer.Write.HandleMetadata.cs`, although a change to a nullable return type for the `ReferenceResolver.GetReference(...)` method would presumably be breaking for implementations. I can contribute an implementation as a PR if this change to the functionality is agreed with by the project.

**Comment by eiriktsarpalis** [comment]
How would you want to be able to control which objects emit id's? Would it be scoped to specific types, using a predicate on the individual object or something else?

**Comment by mihtjel** [comment]
The short form answer to your question is "specific types, but really up to the developer writing the ReferenceResolver to determine." The rest below is just context :D

For the use case prompting this, I have created an Attribute to be used on the key property of any class that is to be referenced - and a custom ReferenceResolver, which then returns an id value of "propertyName: value", resulting in `"$id": "propertyName: value"` in the JSON. This makes some sense in our particular business environment.

For example, a Person might have a number of Contracts, each of which have an ContractId, and one of which is the Primary Contract:
```
{
  "personId": 123,
  "contracts": [
    {
      "$id": "contractId: 1",
      "contractId:" 1,
      "department": "bla bla",
...
    },{
      "$id": "contractId: 2",
      "contractId": 2,
...
    }
  ],
  "primaryContract": {
    "$ref": "contractId: 2",
```

Here, I would use my ReferenceKeyAttribute on the ContractId property of the Contract class, indicating that reference IDs should be generated for these objects.

**Comment by eiriktsarpalis** [comment]
> The short form answer to your question is "specific types, but really up to the developer writing the ReferenceResolver to determine." The rest below is just context :D

Right, but presumably that would require some kind of new API to be added to `ReferenceResolver`, right? Do you envision how that might look?

**Comment by mihtjel** [comment]
> Right, but presumably that would require some kind of new API to be added to `ReferenceResolver`, right? Do you envision how that might look?

So, as far as I can figure out, if `GetReference` in `ReferenceResolver` had `string?` instead of `string` as the return type, the current code (except for debug assertions) should actually do what I want.

`GetReference` is called from here:
https://github.com/dotnet/runtime/blob/0421836895392126ec33486c0f6d1d90fc07b84a/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs#L77

And looking at `WriteMetadataForObject` in the same file, on L28-32, the `"$id"` property is not written if `NewReferenceId` is `null`. 

I believe changing the return type would probably be quite breaking for any existing users of the `ReferenceResolver` class, and I don't have a great design for avoiding this on hand. Adding an `out bool` similar to the current `alreadyExists` may be a more compatible change, but I'm really not that well versed in how to design this sort of API change.

**Comment by mihtjel** [comment]
I'm back in the office now, and just tried it out: My custom ReferenceResolver uses the tagged property if present, and if no such property existed, it previously generated consecutive IDs a'la PreserveReferenceResolver.

I replaced this with `return null!;`, and now I only get metadata on the objects where I have specified an ID property. The only remaining issue for me is that it feels very wrong to return null from a method with a non-nullable return type.

**Comment by eiriktsarpalis** [comment]
Perhaps that suggests that the nullability annotations + documentation could be updated to formally support this.
